bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -685,27 +686,38 @@ bool ReplicaImp::relevantMsgForActiveView(const T *msg) {
   }
 }
 
-bool ReplicaImp::validatePreProcessedResults(const PrePrepareMsg *msg,
-                                             const bftEngine::impl::ReplicasInfo &replicasInfo) {
+bool ReplicaImp::validatePreProcessedResults(const PrePrepareMsg *msg) {
   RequestsIterator reqIter(msg);
   char *requestBody = nullptr;
+  std::vector<std::future<void>> tasks;
+  std::vector<std::optional<std::string>> errors(msg->numberOfRequests());
+  size_t error_id = 0;
+
   while (reqIter.getAndGoToNext(requestBody)) {
     const MessageBase::Header *hdr = (MessageBase::Header *)requestBody;
     if (hdr->msgType == MsgCode::PreProcessResult) {
-      preprocessor::PreProcessResultMsg req((ClientRequestMsgHeader *)requestBody);
-
-      if (auto err = req.validatePreProcessResultSignatures(getReplicaConfig().replicaId, getReplicaConfig().fVal);
-          err) {
-        // trigger view change
-        LOG_INFO(VC_LOG,
-                 "Received PrePrepareMsg containing PreProcessResult with invalid signature: "
-                     << *err << ". Ask to leave view " << getCurrentView());
-        askToLeaveView(ReplicaAsksToLeaveViewMsg::Reason::PrimarySentBadPreProcessResult);
-        return false;
-      }
-    } else {
-      // ClientRequest messages are handled in PrePrepare::validate()
-      continue;
+      tasks.push_back(RequestThreadPool::getThreadPool().async(
+          [&errors, requestBody, error_id](auto replicaId, auto fVal) {
+            preprocessor::PreProcessResultMsg req((ClientRequestMsgHeader *)requestBody);
+            errors[error_id] = req.validatePreProcessResultSignatures(replicaId, fVal);
+          },
+          getReplicaConfig().replicaId,
+          getReplicaConfig().fVal));
+      error_id++;
+    }
+  }
+  for (const auto &t : tasks) {
+    t.wait();
+  }
+  errors.resize(error_id);
+  for (auto err : errors) {
```

Comment:
Amitava, from reading this code, I don't understand how we only check for `f+1` valid signatures, instead of `n`?

Reply:
Thank you @teoparvanov for this comment. As per my understanding, every replica receives f signatures. I will discuss with @tdimitrov and then ans this comment.

Reply:
@teoparvanov , I had a discussion with @tdimitrov . We get only f signatures, because the primary also executes the requests. So the current logic is fine.

Reply:
Hi Amitava, thanks for clarifying!